msprime/likelihood.py

```diff
@@ -0,0 +1,195 @@
+#
+# Copyright (C) 2019 University of Oxford
+#
+# This file is part of msprime.
+#
+# msprime is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# msprime is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with msprime.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Module responsible for computing likelihoods.
+"""
+import math
+import msprime
+import numpy as np
+
+
+def unnormalised_log_mutation_likelihood(arg, theta):
+    # log_likelihood of mutations on a given ARG up to a normalising constant
+    # that depends on the pattern of observed mutations, but not on the ARG
+    # or the mutation rate
+    tables = arg.tables
+    number_of_edges = len(tables.edges)
+    edges_above = {}
+    edge = 0
+    total_material = 0
+    while edge < number_of_edges:
+        child = tables.edges[edge].child
+        right = tables.edges[edge].right
+        left = tables.edges[edge].left
+        parent = tables.edges[edge].parent
+        total_material += ((right - left) *
+                           (tables.nodes[parent].time - tables.nodes[child].time))
+        edges_above[child] = [edge]
+        edge += 1
+        if edge == number_of_edges:
+            break
+        while tables.edges[edge].child == child:
+            right = tables.edges[edge].right
+            left = tables.edges[edge].left
+            parent = tables.edges[edge].parent
+            total_material += ((right - left) *
+                               (tables.nodes[parent].time - tables.nodes[child].time))
+            edges_above[child].append(edge)
+            edge += 1
+            if edge == number_of_edges:
+                break
+    edges_below = {}
+    edge = 0
+    while edge < number_of_edges:
+        parent = tables.edges[edge].parent
+        edges_below[parent] = [edge]
+        edge += 1
+        if edge == number_of_edges:
+            break
+        while tables.edges[edge].parent == parent:
+            edges_below[parent].append(edge)
+            edge += 1
+            if edge == number_of_edges:
+                break
+    number_of_mutations = len(tables.mutations)
+    if theta == 0:
+        if number_of_mutations == 0:
+            ret = 0
+        else:
+            ret = -float("inf")
+    else:
+        ret = (number_of_mutations * math.log(total_material * theta) -
+               total_material * theta)
+    for mut in arg.mutations():
+        mutant_location = tables.sites[mut.site].position
+        ind = 0
```

Comment:
What's ``ind`` for here?

Reply:
A dummy variable for finding the edge in `edges_above[mut.node]` that is ancestral at `tables.sites[mut.site].position` using the loop at line 83. There's probably a more elegant way to do that though.